tsdb/store.go

```diff
@@ -804,6 +804,32 @@ func (s *Store) SetShardEnabled(shardID uint64, enabled bool) error {
 	return nil
 }
 
+// SetShardNewReadersBlocked sets if new readers can access the shard. If blocked
+// is true, the number of reader blocks is incremented and new readers will
+// receive an error instead of shard access. If blocked is false, the number
+// of reader blocks is decremented. If the reader blocks drops to 0, then
+// new readers will be granted access to the shard.
+func (s *Store) SetShardNewReadersBlocked(shardID uint64, blocked bool) error {
+	sh := s.Shard(shardID)
+	if sh == nil {
+		return ErrShardNotFound
+	}
+	return sh.SetNewReadersBlocked(blocked)
+}
+
+// ShardInUse returns true if a shard is in-use (e.g. has active readers).
+// SetShardNewReadersBlocked(id, true) should be called before checking
+// ShardInUse to prevent race conditions where a reader could gain
+// access to the shard immediately after ShardInUse is called.
+func (s *Store) ShardInUse(shardID uint64) (bool, error) {
+	sh := s.Shard(shardID)
+	if sh == nil {
+		return false, ErrShardNotFound
+	}
+	sh.ScheduleFullCompaction()
```

Comment:
Why do you schedule a compaction here?

Reply:
Without this, all the data remains in the WAL, so there are no TSM files and no calls to `TSMReader.Ref`. Since the test requires calls to `TSMReader.Ref`, this is an issue.

I copied this from other tests that also need to flush data from the WAL to TSM files. I'm confused why this test and the others work without waiting for the compaction to finish, but they seem to consistently work.

If the are other better ways of doing this, I would be open to changing it over.

Reply:
Oh, wait a minute. In context I see that this is some mystery copy and paste. Let me go remove it!

Reply:
`ScheduleFullCompaction` now properly called in test code.